openedx/core/djangoapps/schedules/tasks.py

```diff
@@ -3,17 +3,46 @@
 from celery.task import task
 from django.conf import settings
 from django.contrib.sites.models import Site
+from django.core.exceptions import ValidationError
 from django.core.urlresolvers import reverse
+from django.db.utils import DatabaseError
 from django.utils.http import urlquote
 
+from logging import getLogger
+
 from edx_ace import ace
 from edx_ace.message import MessageType, Message
 from edx_ace.recipient import Recipient
 from edx_ace.utils.date import deserialize
+from opaque_keys.edx.keys import CourseKey
 from openedx.core.djangoapps.schedules.models import Schedule, ScheduleConfig
 
 
+log = getLogger(__name__)
+
+
 ROUTING_KEY = getattr(settings, 'ACE_ROUTING_KEY', None)
+KNOWN_RETRY_ERRORS = (  # Errors we expect occasionally that could resolve on retry
+    DatabaseError,
+    ValidationError,
+)
+
+
+@task(bind=True, default_retry_delay=30, routing_key=ROUTING_KEY)
+def update_course_schedules(self, **kwargs):
+    course_key = CourseKey.from_string(kwargs['course_id'])
+    new_start_date = deserialize(kwargs['new_start_date_str'])
+    new_upgrade_deadline = deserialize(kwargs['new_upgrade_deadline_str'])
+
+    try:
+        Schedule.objects.filter(enrollment__course_id=course_key).update(
```

Comment:
Maybe only change `Schedules` that have the previous date set as their start date? That would potentially make this idempotent even if we decide to split it into smaller transactions (to prevent a single huge database write).

Reply:
That's a good idea.  I time-boxed the effort to make this change.  Unfo, the timestamp granularity stored in Mongo is different from that stored in MySQL, and the support in django's ORM for filtering with insensitivity to seconds isn't that great in 1.8 (much better in 1.9).  We can revisit this idea if/when we choose to split the sql update.